### PR TITLE
feat: java language support

### PR DESCRIPTION
**Pull request by vjerci**
## Description
This pr adds basic java support.
It adds string detector, object detector, as well as flow analysis support for java.

This pr as well adds support for integration testing rules with certain compositions.

## Related
https://github.com/Bearer/bearer/issues/894

## Checklist

- [x] I've added test coverage that shows my fix or feature works as expected.
- [x] I've updated or added documentation if required.
- [x] I've included usage information in the description if CLI behavior was updated or added.
- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/) format

### CODE REVIEW

File: new/detector/implementation/java/object/object.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	results, err := detector.classNameQuery.MatchAt(node)
+	if len(results) == 0 || err != nil {
+		return nil, err
+	}
+
+	result := results[0]
+
+	className := result["class_name"].Content()
+
+	properties, err := generic.GetNonVirtualObjects(evaluator, result["right"])
```
Comment (didroe) [inline new/detector/implementation/java/object/object.go]: Does `result["right"]` exist?

File: new/detector/implementation/java/object/object.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	fieldAcessQuery, err := lang.CompileQuery(`(field_access object: (_) @object field: (identifier) @field) @root`)
+	if err != nil {
+		return nil, fmt.Errorf("error compiling call query: %s", err)
+	}
+
+	return &objectDetector{
+		assignmentQuery:    assignmentQuery,
+		classNameQuery:     classNameQuery,
+		classPropertyQuery: classPropertyQuery,
+		fieldAcessQuery:    fieldAcessQuery,
```
Comment (didroe) [inline new/detector/implementation/java/object/object.go]:
```suggestion
		fieldAccessQuery:    fieldAccessQuery,
```

File: new/detector/implementation/java/ruby_test.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package ruby_test
```
Comment (didroe) [inline new/detector/implementation/java/ruby_test.go]:
```suggestion
package java_test
```
(also filename and test names)

File: new/detector/implementation/java/string/string.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+func (detector *stringDetector) Name() string {
+	return "string"
+}
+
+func (detector *stringDetector) DetectAt(
+	node *tree.Node,
+	evaluator types.Evaluator,
+) ([]interface{}, error) {
+	if node.Type() == "string_literal" {
```
Comment (didroe) [inline new/detector/implementation/java/string/string.go:27]: We could also support `+`

File: new/language/implementation/java/java.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+		// public class Main {
+		//
+		// }
+		case "class_body":
+			previousScope := scope
+			scope = implementation.NewScope(previousScope)
+			err := visitChildren()
+			scope = previousScope
+			return err
+			// public class Main {
```
Comment (didroe) [inline new/language/implementation/java/java.go]:
```suggestion
		// public class Main {
```

File: new/language/implementation/java/java.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+func (javaImplementation *javaImplementation) ShouldSkipNode(node *tree.Node) bool {
+	return false
+}
+
+func (*javaImplementation) PatternLeafContentTypes() []string {
+	return []string{
+		// todo: see if type identifier should be removed from here (User user) `User` is type
+		// todo see if `modifier` should be here it's a single token containing `string public final static`
```
Comment (didroe) [inline new/language/implementation/java/java.go]: - Identifier is good I think
- Yes I'd say add `modifier`, though we probably won't use it in practice

**Review by didroe — APPROVED** [review]

File: new/detector/implementation/java/object/object.go (hunk 1 above)
Comment (vjerci) [reply new/detector/implementation/java/object/object.go]: yep

File: new/detector/implementation/java/string/string.go (hunk 4 above)
Comment (vjerci) [reply new/detector/implementation/java/string/string.go:27]: lets see if we need it.

File: new/detector/implementation/java/object/object.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	return objects, nil
+}
+
+func (detector *objectDetector) getClassName(node *tree.Node, evaluator types.Evaluator) ([]interface{}, error) {
+	results, err := detector.classNameQuery.MatchAt(node)
+	if len(results) == 0 || err != nil {
+		return nil, err
+	}
+
+	result := results[0]
```
Comment (didroe) [inline new/detector/implementation/java/object/object.go]: If you're only expecting a single result you can use `MatchOnceAt` instead.

File: new/detector/implementation/java/object/object.go (hunk 1 above)
Comment (didroe) [reply new/detector/implementation/java/object/object.go]: I don't see it here: 
https://github.com/Bearer/bearer/pull/919/files#diff-f2858e7bb9c4cf328b23d5a8c3543c6d51b3819021dddda306e52944026c870bR36-R39

Comment (vjerci) [reply new/detector/implementation/java/object/object.go]: my bad thought this one is about assigmentQuery, will fix it

File: new/detector/implementation/java/object/object.go (hunk 7 above)
Comment (vjerci) [reply new/detector/implementation/java/object/object.go]: shouldn't that method be called `getSingleResult`? and the other one probably `getAllNestedResults`? `matchOnceAt` doesn't seems explicit but implicit.

Comment (vjerci) [reply new/detector/implementation/java/object/object.go]: Fixed, I'm resolving this, and merging, However, I still think that the method naming there is a bit confusing.

**Pull request merged**